minio/minio
# Dynamic usage cache

## Motivation and Context

The current usage cache has a fixed level at which it compacts directories,
meaning everything beneath `bucket/prefix/subprefix` is compacted into one entry.

While this covers a lot of scenarios this is not optimal in all cases, for instance shallow structures that store a lot of objects at a high prefix level or deep structures that store a lot of data with shared prefixes.

The downside will be either a lot of memory usage in case of a shallow structure or high crawl cycle times it is a deep structure.

We change the compaction to be dynamic so it can adapt to the structure of the data.


## Description

 Cache structure and compaction:

 A cache structure will be kept with a tree of usages.
 The cache is a tree structure where each keeps track of its children.

 An uncompacted branch contains a count of the files only directly at the
 branch level, and contains link to children branches or leaves.

 The leaves are "compacted" based on a number of properties.
 A compacted leaf contains the totals of all files beneath it.

 A leaf is only scanned once every dataUsageUpdateDirCycles,
 rarer if the bloom filter for the path is clean and no lifecycles are applied.
 Skipped leaves have their totals transferred from the previous cycle.

 A clean leaf will be included once every healFolderIncludeProb for partial heal scans.
 When selected there is a one in healObjectSelectProb that any object will be chosen for heal scan.

 Compaction happens when either:

 1) The folder (and subfolders) contains less than dataScannerCompactLeastObject objects.
 2) The folder itself contains more than dataScannerCompactAtFolders folders.
 3) The folder only contains objects and no subfolders.

 A bucket root will never be compacted.

 Furthermore if a has more than dataScannerCompactAtChildren recursive children (uncompacted folders)
 the tree will be recursively scanned and the branches with the least number of objects will be
 compacted until the limit is reached.

 This ensures that any branch will never contain an unreasonable amount of other branches,
 and also that small branches with few objects don't take up unreasonable amounts of space.

 Whenever a branch is scanned, it is assumed that it will be un-compacted
 before it hits any of the above limits.
 This will make the branch rebalance itself when scanned if the distribution of objects has changed.

TLDR; With current values: No bucket will ever have more than 10000 child nodes recursively.  No single folder will have more than 2500 child nodes by itself. All subfolders are compacted if they have less than 500 objects in them recursively.

We accumulate the (non deletemarker) version count for paths as well, since we are changing the structure anyway.


## How to test this PR?

Enable server debug to observe scanning progress: `_MINIO_SERVER_DEBUG=on`. Observe crawling while updating files.

## Types of changes
